### PR TITLE
[MODULAR] Crew monitoring console now becomes needy when the crew is.

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request

Alright. So, with the scarcity of handheld crew monitors, combined with people cryoing with gear, taking job slots, not doing their jobs late into the round, and breaking the need to glue yourself to a console otherwise... I have decided to make the Crew Monitor consoles glow a nice red and beep quietly every 10 seconds while there are ongoing medical emergencies (on sensors only).

This should prevent incidents I have seen in the past where admin intervention is require 30+ minutes after someone died on sensors because medical cannot be bothered to do their jobs or respond to suit sensors... command surprised pikachu'ing when they hear someone's been dead 30+.. etc... 

This should attract attention to the crew console when someone is soft crit or worse on it. 

## How This Contributes To The Skyrat Roleplay Experience

It sucks to be a player watching medical fumble around bored when there's people dead on suit sensors. When there are no dedicated paramedics that are amazing at their jobs, this will make medical noticing you FAR more likely. It also makes sense a bit for a system that detects emergencies to also alert you, like medibots do. 

## Media
dreamseeker_zB0q7ZbHi7

https://user-images.githubusercontent.com/77420409/165019077-d82c7e77-84ec-43ff-bb08-131aa5c028ea.mp4

https://user-images.githubusercontent.com/77420409/165019092-4f4ed12f-dbd1-49ff-b5e9-016e8680a8f5.mp4

## Changelog

:cl:
add: Suit sensors console beeps and glows for the injured.
/:cl:

### CODE REVIEW

**Comment by Zergspower** [comment]
@nevimer I was about to PR the fix hah, but so we dont have conflicts can you fix the last column not collapsing , the val is there but for some reason didnt apply the collapsing mod.

**Comment by nevimer** [comment]
Oh, this PR touches none of what you did! This is entirely an override. @Zergspower

**Comment by Zergspower** [comment]
weh I wanted to be lazy

File: modular_skyrat/modules/crewconsole/glow.dm [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#define SENSORS_UPDATE_PERIOD 10 SECONDS //Why is this not a global define, why do I have to define it again
```
Comment (Zonespace27) [inline modular_skyrat/modules/crewconsole/glow.dm:1]: undef

Comment (nevimer) [reply modular_skyrat/modules/crewconsole/glow.dm:1]: done

**Comment by theselfish** [comment]
Make the screen glow red.

**Comment by Zergspower** [comment]
> Make the screen glow red.

my concern there would be 1 assistant dead with their sensors on binary off in some corner where they're never found lol

**Comment by Gandalf2k15** [comment]
This is actually a really big powercreep for this console, traitors and such rely on people not actually paying attention to the console, so it's a large balance change if it makes a ping every time someone with sensors dies.

**Comment by Higgin** [comment]
code and feature wise what is there to say if it works? i'd like to be able to multitool the board or configure the feature on the console itself (maybe emagging it allows it to be permanently disabled?) i keep a handheld in my pocket and up for probably 80% of the round when playing any role in medical where i can get one - the beeps from a console for the unknown dead on the assault ops ship would get very obnoxious if they're that loud at that distance - even making it something like every 1-3 minutes would be a big improvement. 


on the server side:
if this goes through there needs to be more explicit permission given for hiding bodies and stripping/turning off the sensors of the dead

if you want to kill somebody and have more than a ~30s-1m window before cyborgs, paramedics, or doctors with attitude are kicking down your door, combined with the time your victim might have access to their headset, this almost makes using the sleepy pen or EMP a necessity. you're channeled into a much smaller set of options to avoid having to softcrit people first and set this off because a beep telling any one of a number of people near a sensor monitor 'hey i should be doing my job' makes it immediately riskier that they will - i shut down a lot of people just by watching sensors already

also, honestly, sometimes you **do** get messed up and get out for the rest of the round - we should be norming that and allowing rounds to end earlier through votes or spiraling conflict (enabled by more frequent midrounds that folks who've been RR'd - even soft RR'd folks - can use to rejoin.) if being dead sucks, but people generally have fun doing things where there actually is a risk of death, injury, or nonrecovery, lessening the cost of dying in OOC time has to be balanced against having any death matter or be worth inflicting on somebody - i think where we're at, this balance is better struck through giving people more opportunities to rejoin the game and participate in conflict as blank slate midrounds rather than the character they joined in as. and allowing rounds where people get RR'd to move towards ending earlier and more dramatically, or just towards ending earlier in general, is a good way to lessen the cost without saying that revival after death should be guaranteed (a position this PR seems to advance)

**Comment by YakumoChen** [comment]
Agree with Gandalf, this basically hazes traitors out of killing entirely or makes EMP an absolute requirement for almost any combat whatsoever, instead of being merely more likely when medbay or the warden is bored.

A far more elegant solution is for this to activate ONLY for the dead, and preferably after a time minimum is reached, like 5 minutes.

**Comment by nevimer** [comment]
You all are missing the point - this is to help medbay in general. I guarantee you this changes nothing as far as balance cares. If you have any of the jobs watching suit sensors that should be - CMO, Captain, Paramedic, Orderly, Warden, Medical Doctor, any command or medical role really.. Even the HOS.. NOTHING CHANGES. This isn't even remotely as impactful as the warden dispatch console. 

Y'all. We have atmos alarms, engineering alarms, that you can watch like a hawk to catch blobs, vault raids, etc super easy the moment they occur. ALL this power existed already, and I know it does cause the moment I get injured already I am stormed by sec and medical, reliably. 

On understaffed medbay rounds? Or rounds where the department is not as robust, the console will emit a soft glow and emit a gentle beep every ten seconds. Why? Because if it lasts that long, medical isn't paying attention. 

This notably comes to mind on maps like blueshift or icebox where people die exploring and medical needs a LOT of prompting. 

Medical just often times needs a lot of prompting yo do their jobs. 

I can move it to 60 seconds per update. The default TG decided for suit sensors updates were 10 seconds, however. I can even move it to hard crit, or dead. The end goal is to keep people playing.

**Comment by YakumoChen** [comment]
You're missing the point that suit sensors rely on that very action of people doing their job. As often is the case, to the benefit of antagonists, the primary driver of rounds, there are not enough people paying attention.

If the console goes ballistic within seconds of someone dying forcing everyone with a console to pay attention, you remove all ability for any round drivers to do their job without extremely specific equipment.

Players aren't entitled to being immediately brought back into the round on death, unless their corpse is left in front of medbay or somewhere extremely obvious.

**Comment by YakumoChen** [comment]
It's a hypocritical point to make anyway that "this changes nothing"

If people were paying enough attention in the first place this PR wouldn't have been made.

**Comment by nevimer** [comment]
> You're missing the point that suit sensors rely on that very action of people doing their job. As often is the case, to the benefit of antagonists, the primary driver of rounds, there are not enough people paying attention.
> 
> If the console goes ballistic within seconds of someone dying forcing everyone with a console to pay attention, you remove all ability for any round drivers to do their job without extremely specific equipment.
> 
> Players aren't entitled to being immediately brought back into the round on death, unless their corpse is left in front of medbay or somewhere extremely obvious.

You would THINK this is true, but medical can and will leave heads laying on tables, in their lobby, on the floor, etc, for very long periods of time unless pointed out to them. This doesn't solve visual clarity issues, but it targets the specific problem of medical not doing their jobs - the only person being punished for that is those who have no control over that. 

The root of the issue is medical needs prompting to do their job, the skill level of the medbay varies wildly round to round. Some rounds, medical just morgues reviveable bodies. I'll add a delay to it of 2 minutes, make it check for souls/DNR status so we don't get humonkeys going. I still want the hawks watching it to be rewarded, but I also don't want people to be out of the round due to medbay's "skill issues". It's just not fair to the player to watch medical ERPing and complaining about greenshifts, meanwhile people are actually dead on sensors - just Pandora's box mentality.

**Comment by YakumoChen** [comment]
> You would THINK this is true, but medical can and will leave heads laying on tables, in their lobby, on the floor, etc, for very long periods of time unless pointed out to them. This doesn't solve visual clarity issues, but it targets the specific problem of medical not doing their jobs - the only person being punished for that is those who have no control over that.

I'm of the opinion this serves largely as a crutch which impedes antagonists more than it does help medbay. It sets a bad precedent for jobs and, As Gandalf said, it represents a large balance change.

More than that, it's going to go off constantly anytime someone dies, whether they're in maint or already on a medbay table. 
Even more than that, there are monitoring consoles in many other places. The bridge and warden have one, this will lead to way more console hawking and annoyance than is needed for a round.

Death is more meaningless with additions like these, largely, players will play to the crutch of expecting to be revived even more than they do. There are respawns and ghost roles, players are not forbidden from re entering the round on death.

> The root of the issue is medical needs prompting to do their job, the skill level of the medbay varies wildly round to round. Some rounds, medical just morgues reviveable bodies. I'll add a delay to it of 2 minutes, make it check for souls/DNR status so we don't get humonkeys going. I still want the hawks watching it to be rewarded, but I also don't want people to be out of the round due to medbay's "skill issues". It's just not fair to the player to watch medical ERPing and complaining about greenshifts, meanwhile people are actually dead on sensors - just Pandora's box mentality

The worst precedent this sets is that it implies every department needs an immutable crutch. Not every department functions perfectly every round. That's just the way of things. Sometimes Security gets wiped out or are too dumb to deal with antags. Sometimes the supermatter blows up. Sometimes medbay is slow or understaffed and dead people are harder to revive. Do we need automated answers to prop up every failing of a department? Especially one that so severely hinders antagonists? I'm not so sure.

**Comment by YakumoChen** [comment]
It should go without saying that the biggest benefactor of this pr is security. When the console sends a loud and annoying alert to the medbay bridge, and brig, because an officer died, what do you expect the response to be? Its going to be the brunt of all three departments rushing to the location.

**Comment by Deek-Za** [comment]
It's a cool PR but I would not necessarily agree with the idea.  Think the medical team should be responsible for keeping watch and if need-be the CMO can assign someone to be on monitor duty.
I think what potentially could be a better solution is maybe a couple extra monitors dispersed through medbay? And not the hard ground computer types. More like the ones that hang off the wall types. 
Heck even my actual Emergency Department I work at has those hanging down from the ceiling, with vital information for patients hooked up to the machines.

Having them around in more areas might make it a tad bit more convenient since not every map's medbay has a proper front desk to interact with crew while keepin' an eye on things.

Plus the red glow is kind of an eye sore.

**Comment by SpookyTheFox** [comment]
Rather than convert the existing consoles into harm alarms, It would be good to explore how we got into the situation that makes a code change potentially necessary. I think there's been a long term decline in diligence and competency in the medical department, the result of which has been long revive times, and dead crew being missed or even ignored.

1. Watching the crew monitor is pretty boring, traditionally borgs (and only borgs) bothered to watch it, for external reasons. They need the crew monitor back.
2. The crew monitor also doesn't put the most injured crew at the top any more... why was this change made? There's a lot of spacemans on the station, you have to constantly scroll up and down to find the injuries without it auto sorting.
3. Further, medical has little incentive to stay focused. Changes to research nodes, item removals, and administrator opinions have dramatically cut down on implant installs, which has traditionally been busy work and a skill builder for medbay. With nothing to do in medbay, the doctors aren't in the mindset of being responsive to increases in activity that happen part way through the round.

TL;DR: I think a larger conversation needs to be had on what changes in code have led to changes in player behavior, that have made the existing setup not work anymore. Then, decide if that trend will continue and you need to modify the console in hopes that it changes the way players act, or if you should consider undoing some of the changes that led you here.

**Comment by nevimer** [comment]
> Rather than convert the existing consoles into harm alarms, It would be good to explore how we got into the situation that makes a code change potentially necessary. I think there's been a long term decline in diligence and competency in the medical department, the result of which has been long revive times, and dead crew being missed or even ignored.
> 
> 1. Watching the crew monitor is pretty boring, traditionally borgs (and only borgs) bothered to watch it, for external reasons. They need the crew monitor back.
> 2. The crew monitor also doesn't put the most injured crew at the top any more... why was this change made? There's a lot of spacemans on the station, you have to constantly scroll up and down to find the injuries without it auto sorting.
> 3. Further, medical has little incentive to stay focused. Changes to research nodes, item removals, and administrator opinions have dramatically cut down on implant installs, which has traditionally been busy work and a skill builder for medbay. With nothing to do in medbay, the doctors aren't in the mindset of being responsive to increases in activity that happen part way through the round.
> 
> TL;DR: I think a larger conversation needs to be had on what changes in code have led to changes in player behavior, that have made the existing setup not work anymore. Then, decide if that trend will continue and you need to modify the console in hopes that it changes the way players act, or if you should consider undoing some of the changes that led you here.

Yeah, we have lots of issues - and my primary issue I've noticed is medical doesn't do their job unless you literally tell them to do it - it's not medical bring in bodies most of the time - its security or others. 

Medical in general will not notice corpses, nor will common players even, unless specifically told. 

See screenshot examples: 

zRdHLPbnKB

dreamseeker_HiqR4bWEkg
dreamseeker_6SVbRw4XqS
yaGYxtGl8x

**Comment by Deek-Za** [comment]
Could the crew monitor maybe also be made into a tablet(or now modular PDA) app?.
Maybe that'd help it work better. I mean as regular Engi department player I'm pretty much almost never separated from the canary alarm on my screen to help rectify solutions. Admittedly I'm a first responder more often than most of my engineering brethren. But maybe giving the potential for other's to stick a crew monitor on their screen and respond or alert the Parameds to an immediate emergency might go for a distance. Cause then we're saving folk but also letting it be player response based over mechanical stuff. Also might separate the dedicated life savers from the slackers.

**Comment by theselfish** [comment]
Skyrat players are morons. Congratulations on figuring that out. I'm glad to be the first to welcome you to our server.

Giving antags a very small time frame after they kill someone isn't the answer.
Something like the ...I forget the name, but Virgo had this ability to ping Medical after five minutes. Maybe implement something like that.

**Comment by nevimer** [comment]
> 



> Skyrat players are morons. Congratulations on figuring that out. I'm glad to be the first to welcome you to our server.
> 
> Giving antags a very small time frame after they kill someone isn't the answer. Something like the ...I forget the name, but Virgo had this ability to ping Medical after five minutes. Maybe implement something like that.

I know how skyrat players work, hence this PR. No need to insult them, they're not all mega-robust 24/7 gamers. This is something that's done on baystation, too. We shouldn't "balance" around 24/7 gaming, instead, enjoyment. I don't consider this a balance change.

Their system is called transcore, by the way. That'd be going a step further, which I am willing to do. In the current commit, the consoles are 1% chance every 10 seconds to speak up about suit sensors on med comms.

**Comment by YakumoChen** [comment]
> This is something that's done on baystation, too. We shouldn't "balance" around 24/7 gaming, instead, enjoyment. I don't consider this a balance change.

It absolutely is a balance change. It is a direct benefit to security and a direct threat to antagonists, who already are extremely discouraged from being violent with gunsec back in force. This will make playing roles like Heretic who lack access to EMPs nigh impossible and EMP a requirement to do anything without the threat of 10 well armed secoffs being shoved down their throat.

**Comment by nevimer** [comment]
> > This is something that's done on baystation, too. We shouldn't "balance" around 24/7 gaming, instead, enjoyment. I don't consider this a balance change.
> 
> It absolutely is a balance change. It is a direct benefit to security and a direct threat to antagonists, who already are extremely discouraged from being violent with gunsec back in force. This will make playing roles like Heretic who lack access to EMPs nigh impossible and EMP a requirement to do anything without the threat of 10 well armed secoffs being shoved down their throat.

You're talking to a person who thinks EMP's are awful game design, too, so...

**Comment by YakumoChen** [comment]
Do you think traitors should be forced to expect a security response every time they crit an officer, within 0-10 seconds of hitting crit?

**Comment by nevimer** [comment]
> Do you think traitors should be forced to expect a security response every time they crit an officer, within 0-10 seconds of hitting crit?

Sorta, yeah. Depending on how loud they are, and where, and the method they use. This game takes place ins 2562 or something. Space age - we should have some form of automated alerts. We have them for atmos - which already outs traitors, blobs, etc, instantly. Especially to cyborgs.

**Comment by YakumoChen** [comment]
I might agree with you if we actually enforced policy that not every threat needs a response of the entire department but as it is now you're encouraging antagonists to be forced to fight up to 10 well-armed security every time they do anything vaguely violent.

Do you not see how this completely discourages anyone from playing antagonist?

**Comment by nevimer** [comment]
> 



> I might agree with you if we actually enforced policy that not every threat needs a response of the entire department but as it is now you're encouraging antagonists to be forced to fight up to 10 well-armed security every time they do anything vaguely violent.
> 
> Do you not see how this completely discourages anyone from playing antagonist?

I see.. Well. What would make you happy? Only alarm on deaths instead of crit? I included crit for reasons of people being stuck in softcrit perpetually.

**Comment by YakumoChen** [comment]
Pretty sure Spooky had it right the first time. The problem is not that nobody is checking the monitor, the problems are with how the monitor currently doesn't do a sufficiently good job at being readable, and that medbay is a very boring job due to policy and code changes. Recent trends that have discouraged antagonists to be violent (LIKE THIS PR) have only left medbay with even less to do as there is not sufficient conflict to keep people coming to the department.

**Comment by YakumoChen** [comment]
> > 
> 
> > I might agree with you if we actually enforced policy that not every threat needs a response of the entire department but as it is now you're encouraging antagonists to be forced to fight up to 10 well-armed security every time they do anything vaguely violent.
> > Do you not see how this completely discourages anyone from playing antagonist?
> 
> I see.. Well. What would make you happy? Only alarm on deaths instead of crit? I included crit for reasons of people being stuck in softcrit perpetually.

My very first comment highlights how I believe the PR would work 'best'. It should only count for deaths, and it should only count deaths for those who have been dead for at least 5 minutes. 

Preferably it should ONLY be alerting medical staff. I think the visual annoyances are not needed and only encourage non-medical staff to view it. It should be like medibots, only alerting with the radio. This is something you've already coded in so you clearly accounted for it.

PRs like this trying to tackle a problem _with major balance implications_ should start with very small changes. Radio alerts on long-dead corpses is very reasonable and accomplishes the goal of the PR. If it somehow is not sufficient, escalate the effects to stuff like visuals in a different PR.

**Comment by GoldenAlpharex** [comment]
I think that this looks to be a fix for a symptom, rather than fixing the root of the issue.

There's much better ways to achieve a similar result, where it makes it easier to notice there's injured/dead people on the sensors. Here's a few examples:
- There could be a filtering system, where you could, say, press a button or select an option from a dropdown, and it'd only display those matching that general health status (all, healthy, injured, critical, dead).
- You could have an optional toggle on the menu that displays all the dead at the top, or it could display people from the most to the least injured, based on total damage, while conserving the current sorting order for those that aren't injured (which I think would be nice if it was done in, like, two "sections", where you'd have a section at the top with all the injured, and a section at the bottom with all those that aren't injured.)
- You could look into using an animated icon for those that are heavily injured or dead, which would instantly catch your attention when looking at such lists, because your peripheric vision is still very good at detecting movements, compared to changes in color.

I really don't agree with the PR in its current state, for sure. This is absolutely not the right way to tackle the issue.

**Comment by Gandalf2k15** [comment]
After thought, unfortunately in this case, we probably won't be going with this as it's just too impactful outside of it's intended purpose 👍
